tests/extra_source_and_tests/user_2d_T_shaped_pipe_VIPO_shell/T_shaped_pipe_VIPO_shell.cpp

```diff
@@ -0,0 +1,554 @@
+    Real dt_s = 0.0; /**< Default acoustic time step sizes for solid. */
+    //----------------------------------------------------------------------
+    //	Statistics for CPU time
+    //----------------------------------------------------------------------
+    TickCount t1 = TickCount::now();
+    TimeInterval interval;
+    TimeInterval interval_computing_time_step;
+    TimeInterval interval_computing_pressure_relaxation;
+    TimeInterval interval_updating_configuration;
+    TickCount time_instance;
+    //----------------------------------------------------------------------
+    //	First output before the main loop.
+    //----------------------------------------------------------------------
+    body_states_recording.writeToFile();
+    //----------------------------------------------------------------------------------------------------
+    //	Main loop starts here.
+    //----------------------------------------------------------------------------------------------------
+    while (GlobalStaticVariables::physical_time_ < end_time)
+    {
+        Real integration_time = 0.0;
+        /** Integrate time (loop) until the next output time. */
+        while (integration_time < Output_Time)
+        {
+            time_instance = TickCount::now();
+            apply_gravity_force.exec();
+            Real Dt = get_fluid_advection_time_step_size.exec();
+            //std::cout << "Dt = " << Dt << std::endl;
+            update_fluid_density.exec();
+
+            viscous_acceleration.exec();
+            transport_velocity_correction.exec();
+            /** FSI for viscous force. */
+            viscous_force_on_shell.exec();
+
+            interval_computing_time_step += TickCount::now() - time_instance;
+            time_instance = TickCount::now();
+            Real relaxation_time = 0.0;
+            while (relaxation_time < Dt)
+            {
+                dt = SMIN(get_fluid_time_step_size.exec(), Dt);
+                //std::cout << "dt = " << dt << std::endl;
+
+                pressure_relaxation.exec(dt);
+                /** FSI for pressure force. */
+                pressure_force_on_shell.exec();
+
+                kernel_summation.exec();
+                left_inflow_pressure_condition.exec(dt);
+                up_inflow_pressure_condition.exec(dt);
+                down_inflow_pressure_condition.exec(dt);
+                inflow_velocity_condition.exec();
+
+                density_relaxation.exec(dt);
+
+                Real dt_s_sum = 0.0;
+                average_velocity_and_acceleration.initialize_displacement_.exec();
+                while (dt_s_sum < dt)
+                {
+                    dt_s = shell_time_step_size.exec();
+                    if (dt - dt_s_sum < dt_s)
+                        dt_s = dt - dt_s_sum;
+                    shell_stress_relaxation_first.exec(dt_s);
+
+                    constrain_holder.exec(dt_s);
```

Comment:
You can add damping here if strong shell instability is observed. It's not supposed to be used for dynamic problem, but I found that shell can have strong hourglass effects without using damping

Reply:
Thanks for this advice.